File: Graph-General/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        # Extract rows and columns
        ROWS = len(grid)
        COLS = len(grid[0])
        seen = set()  # hashset to keep visited nodes
        islands = 0  # No of islands

        def bfs(r, c):
            dirs = [(-1, 0), (0, -1), (0, 1), (1, 0)]  # All 4 directions
            q = deque()  # deque to store coordinates of nodes to be visited
            q.append((r, c))  # append current coors to the queue
            seen.add((r, c))  # mark current coordinates as seen

            while q:
                row, col = q.popleft()  # Pop coordinates for deque

                # For each direction
                for dr, dc in dirs:
                    r, c = row + dr, col + dc  # Compute new coordinates

                    # If new coordinates are not in bounds or coordinates have been visited or the node is not an island
                    if (
                        r not in range(ROWS)
                        or c not in range(COLS)
                        or ((r, c) in seen)
                        or grid[r][c] != "1"
                    ):
                        continue  # Skip

                    # Else append to the queue and the hashmap
                    q.append((r, c))
                    seen.add((r, c))

        # Main loop - iterate over every element and if its an island perform bfs
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == "1" and (r, c) not in seen:
                    bfs(r, c)
                    islands += 1  # increment island number

        return islands

File: Graph-General/number-of-islands.cpp
/*
 * Other alternatives for faster solutions:
 * 1. Modify the grid to set "seen" nodes to 0. This way using std::set is redundant.
 * 2. Use unordered_set with custom hashing technique (becasue unordered_set cant accept std::pair as a hash. Instead of that for ex. set unordered_set<int> and store (row * col) + col).
 */
class Solution {
    std::set<std::pair<int, int>> seen = {};

public:
    void bfs(const int &r, const int &c, const size_t &rows, const size_t &cols, const vector<vector<char>>& grid)
    {
        std::queue<std::pair<int, int>> q = {};
        q.push(make_pair(r, c));

        std::vector<std::pair<int, int>> dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        seen.insert(std::make_pair(r, c));

        while(!q.empty())
        {
            auto [row, col] = q.front();
            q.pop();

            for(const auto& [dr, dc] : dirs)
            {
                const int new_row = row + dr;
                const int new_col = col + dc;

                if(new_row < 0 || new_row >= rows || new_col < 0 || new_col >= cols || seen.contains(make_pair(new_row, new_col)) || grid[new_row][new_col] != '1')
                    continue;

                seen.insert(std::make_pair(new_row, new_col));
                q.push({new_row, new_col});
            }
        }
    }

    int numIslands(vector<vector<char>>& grid) {
        size_t rows = grid.size();
        size_t cols = grid[0].size();
        int islands = 0;

        for(int row = 0 ; row < rows ; ++row)
        {
            for(int col = 0 ; col < cols ; ++col)
            {
                if(grid[row][col] == '1' && !seen.contains(std::make_pair(row, col)))
                {
                    bfs(row, col, rows, cols, grid);
                    islands++;
                }
            }
        }
        return islands;
    }
};

